### PR TITLE
Add nibble and transparency support to BltComplexCopy.

### PR DESCRIPTION
**Pull request by bcampbell**
My experiment at using the `action` param as a bitfield to allow nibble-level access to src and target data, and also to add transparency support.
It's a little silly - I think most of the combinations wouldn't be too useful, but hey.
I've only really tested the nibble-pair -> byte blitting, so there may be a couple of mistakes in the others.
Just for my own amusement really - I'm not really expecting you to merge it! But if you do want it, let me know and I'll tidy it up and update the documentation first.

### CODE REVIEW

**Comment by bcampbell** [comment]
New version. It's  a bit clunky - all the combinations of actions and src/target formats tend to lead to an explosion of line-copy routines!
Could be less code using C++ templates. But it'd be less readable :-)
But I'm pretty happy it covers most use-cases I'd expect.
Doesn't handle overlapping blits - would have to do a whole bunch of reverse-copy routines for that, so can't really use it to scroll the screen, say.
The documentation is a bit of a mess in the final pdf. `makedispatch.py`  strips out a lot of formatting. Probably would be nicer to just treat the DOCUMENT chunks as verbatim markdown and insert them as-is. But can sort that out later.
Also, I haven't (yet) done any example code. I plan to  do so when I get a chance!

**Comment by bcampbell** [comment]
Oh - I forgot about the BASIC side. Are there any changes I need to make there?

File: firmware/common/include/interface/blitter.h [hunk 1]
```diff
@@ -14,14 +14,30 @@
...
-	uint16_t 	address;
-	uint8_t 	page,_filler;
-	uint32_t  	stepSize;
-	uint32_t  	stepOffset;
-	uint32_t  	stepCount;
+	uint16_t	address;
+	uint8_t		page;
+	uint8_t		format;     // one of BLTFMT_*
+	uint8_t		k;			// mask in src, solid value in target.
+	uint8_t  	width;		// Size of line, in units (ignored on target).
```
Comment (bocianu) [inline firmware/common/include/interface/blitter.h:21]: Using byte for the width, limits area ro 255 bytes of width. and height is word, where our vertical screen width is 240. I suggest to swap data types for those two values.

File: firmware/common/include/interface/blitter.h [hunk 2]
```diff
@@ -14,14 +14,30 @@
...
+#define BLTFMT_LOW  4		// Low nibble (target only)
+
+
 uint8_t BLTSimpleCopy(uint8_t pageFrom,uint16_t addressFrom, uint8_t pageTo, uint16_t addressTo, uint16_t transferSize);
+
+// Values for BLTComplexCopy action param
+#define BLTACT_COPY 0		// Straight rectangle copy
+#define BLTACT_MASK 1		// Copy, but only where src != srcarea.v (transparent value)
+#define BLTACT_SOLID 2		// Fill with destarea.k (solid value), but only where src != srcarea.k (transparent value)
+
```
Comment (bocianu) [inline firmware/common/include/interface/blitter.h:40]: I think adding BLTACT_XOR 3 might be useful, where dest value is xored with current memory content.

Comment (bcampbell) [reply firmware/common/include/interface/blitter.h:40]: I can definitely see the appeal of a `target = target XOR src` operation, but didn't quite understand the other one. Did you mean xor the target with a constant value?

Comment (paulscottrobson) [reply firmware/common/include/interface/blitter.h:40]: There's no BASIC interface as yet.

**Comment by paulscottrobson** [comment]
My thoughts on this are that I should suggest that you change the blitter interface in any way you think best, as you clearly have more and better ideas than me.  

If this involves leaving it as it is now, that's fine, but I think at this point, even though it's a breaking change, it could be replaced entirely or significantly. 

Worst Case Scenario is the "new blitter" implements the simpler functionality of the old blitter for backward compatibility. This has been done before (with the gamepad for example)

So I haven't approved this PR, not because of the PR itself, but so that you can change it completely if you think it best.

**Comment by bcampbell** [comment]
Haven't had a chance to come back to this yet, but I should be able to over the weekend.

I've got a couple of "approach" questions to put to everyone:

1) Should it stay as a memory-move operation that just happens to be useful for getting images into the framebuffer, or should it explicitly be a image-drawing operation? The difference would be adding support for clipping, say, and taking x,y destination coordinates rather than an address. Or should be a separate API call which is image-oriented? (and probably calls this one after applying some clipping)?

2) In general, should I think of the framebuffer as a straight 8 bits/pixel, 256-entry palette thing, or more like nibble pairs, where one nibble overrides the other to (kind of ) support dual layers (eg background + sprites)? The nibble-pair-override approach can be done using the palette - I get there's no specific support required on the "hardware" side. But I'm curious to know if there's a particular mental model people have in mind for the framebuffer...

**Comment by paulscottrobson** [comment]
1) Can be both. I'd be inclined perhaps towards your second idea, keep the current one more or less as is, or even return it to the original, which just was a fast copier, then the second can be image orientated.

2) Ideally both. The normal "game use" is to update the background - the lower 4 bits of the pixel with drawing and leave the upper 4 bits alone, so there is an unpack facility there, potentially. However, there are already API calls to draw tiles and sprites on the background, so that may be duplication.

**Comment by bcampbell** [comment]
New version tweaks the blitarea struct layout a little (all the fields unused in the target area have been moved to the end of the struct. In theory, the 6502 side could use a smaller memory area for the dest area struct...)

Also adds a new BltImage() function (blitter function 4), which should be a little simpler for common graphics cases, and performs clipping.

**Comment by bcampbell** [comment]
I also wrote a little test program (with llvm-mos) to help me try out all the combinations:
https://gist.github.com/bcampbell/84c598eb11a75b2c5dc51f8fe82b0d5f

**Comment by bcampbell** [comment]
I haven't added any XOR actions, but that could be done later without breaking the API.

**Review by paulscottrobson — APPROVED** [review]

**Comment by paulscottrobson** [comment]
Build isn't working. The fix looks straightforward, but I'd like you to take a look :)

**Pull request merged**